server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java

```diff
@@ -360,6 +361,10 @@ public Query nullValueQuery() {
 
     public abstract Query existsQuery(QueryShardContext context);
 
+    public MatchQuery matchQuery(QueryShardContext context, String analyzer, int slop) {
```

Comment:
Can this be changed to `analyzePhrase(QueryShardContext context, String text, Analyzer analyzer, int slop)` ? 
It seems to much to allow this override so I'd prefer something more focus on the real purpose.

Reply:
or even further: `phraseQuery(BytesRef[] terms, int[] positions)`, ie. only a factory for phrase queries and still leave analysis up to the query parser

Reply:
+1, I think the `FixedShingleFilter` should be able to produce the shingles from the `terms` and `positions` even if there are gaps in the phrase query (produced by stop filter).

Reply:
@jimczi I would be cautious about that. This optimization should return the same hits as a phrase query on the original field. On the regular field, a query for `"the fox"` would be parsed as a term query on `fox`, so it feels wrong to me that we run it as a term query on `_ fox` on the shingle field, as it would have different matches. I'd rather like to document that this option can't optimize queries with stop words.

Reply:
Thanks for clarifying. I agree this is trappy so +1 to disable this optim when gaps are detected and to add documentation about this limitation.

Reply:
We might not want the latter, since the phrase query on the regular field would match everything that has `fox` 2 positions after `Reynard`, while a phrase query for `"reynard _" "_ fox"` adds the additional constraint that there is no token between them. For instance "Reynard cunning fox" would match on the regular field but not on the shingle field.

Even though this might better fit the original intention of the user, I'd like this feature to be as transparent as possible and not change the set of matches for a given query?